pyproject.toml

```diff
@@ -1,7 +1,66 @@
 # Copyright 2022 Canonical Ltd.
 # See LICENSE file for licensing details.
 
-# Testing tools configuration
+[tool.poetry]
+# Charm is not packed as a standard Python package; this information is not used
+name = "charm"
+version = "0.1.0"
+description = ""
+authors = []
+
+[tool.poetry.dependencies]
+python = "^3.10"
+ops = "^2.4.1"
+tenacity = "^8.2.2"
+boto3 = "^1.28.7"
```

Comment:
Caret pins mean less renovate PRs, but there's no real guarantee that minor version changes won't break things. We'll have to pick our poison here, either have to vet more PRs or accept that dependencies can break.

Reply:
If so, why not just pin the exact versions?  That way local and CI builds will all pull a known working version and we wait for renovate to move the pin. Otherwise bulids happening before the renovate schedule will still potentially face issues.